pkg/controller/genericwebhookconfiguration/controller.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package genericwebhookconfiguration
+
+import (
+	"bytes"
+	"fmt"
+
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/kubernetes-sigs/kubebuilder/pkg/webhook/certprovisioner"
+)
+
+type GenericWebhookConfigurationController struct {
+	// kubernetesClientSet is a clientset to talk to Kuberntes apis
+	KubernetesClientSet kubernetes.Interface
+	// kubernetesInformers contains a Kubernetes informers factory
+	KubernetesInformers informers.SharedInformerFactory
+
+	// CertsHandlerFactory knows how to instantiate a CertsHandler.
+	CertsHandlerFactory CertsHandlerFactory
+	// GetCertProvisioner know how to get a CertProvisioner given the common name.
+	GetCertProvisioner func(commonName string) (certprovisioner.CertProvisioner, error)
```

Comment:
> The above two sounds very similar :)

CertsHandler knows how to read and write the certs. For example: using the secret for running on user node or using local fs for running on master node.

GetCertProvisioner knows how to provision the cert. For example: self-signed cert or k8s CA signed.